test/hotspot/jtreg/compiler/inlining/TestSubtypeCheckTypeInfo.java

```diff
@@ -0,0 +1,433 @@
+        if (lateInlineInstanceOfCondPost(a, cond)) {
+            a.m();
+        }
+    }
+
+    /* =========================================================== */
+
+    @InlineFailure
+    // Inlining _isInstance on constant Class compiler/inlining/TestSubtypeCheckTypeInfo$B
+    // @ 2   compiler.inlining.TestSubtypeCheckTypeInfo::lateInlineIsInstanceCondPre (19 bytes)   inline (hot)   late inline succeeded
+    //   @ 7   java.lang.Class::isInstance (0 bytes)   (intrinsic)
+    // @ 9   compiler.inlining.TestSubtypeCheckTypeInfo$A::m (0 bytes)   failed to inline: virtual call
+    static void testIsInstanceCondLate(A a, boolean cond) {
+        if (lateInlineIsInstanceCondPre(a, true)) {
+            a.m();
+        }
+    }
+
+    /* =========================================================== */
+
+    @InlineFailure
+    // Inlining _isInstance on constant Class compiler/inlining/TestSubtypeCheckTypeInfo$B
+    // @ 2   compiler.inlining.TestSubtypeCheckTypeInfo::lateInlineIsInstanceCondPre (19 bytes)   inline (hot)   late inline succeeded
+    //   @ 7   java.lang.Class::isInstance (0 bytes)   (intrinsic)
+    // @ 9   compiler.inlining.TestSubtypeCheckTypeInfo$A::m (0 bytes)   failed to inline: virtual call
+    static void testIsInstanceCondLatePre(A a, boolean cond) {
+        if (lateInlineIsInstanceCondPre(a, cond)) {
+            a.m();
+        }
+    }
+
+    /* =========================================================== */
+
+    @InlineFailure
+    // Inlining _isInstance on constant Class compiler/inlining/TestSubtypeCheckTypeInfo$B
+    // @ 2   compiler.inlining.TestSubtypeCheckTypeInfo::lateInlineIsInstanceCondPost (19 bytes)   inline (hot)   late inline succeeded
+    //   @ 3   java.lang.Class::isInstance (0 bytes)   (intrinsic)
+    // @ 9   compiler.inlining.TestSubtypeCheckTypeInfo$A::m (0 bytes)   failed to inline: virtual call
+    static void testIsInstanceCondLatePost(A a, boolean cond) {
+        if (lateInlineIsInstanceCondPost(a, cond)) {
+            a.m();
+        }
+    }
+
+    /* =========================================================== */
+
+    static abstract class A {
+        public abstract void m();
+    }
+    static abstract class B extends A {
+        public void m() {}
+    }
+
+    static class C extends A {
+        public void m() {}
+    }
+
+    static void lateInline(A o, boolean cond) {
+        if (cond) {
+            o.m();
+        }
+    }
+
+    static boolean lateInlineInstanceOfCondPre(A o, boolean cond) {
+        return cond && (o instanceof B);
+    }
+
+    static boolean lateInlineInstanceOfCondPost(A o, boolean cond) {
+        return (o instanceof B) && cond;
+    }
+
+    static boolean lateInlineIsInstanceCondPre(A o, boolean cond) {
+        return cond && B.class.isInstance(o);
+    }
+    static boolean lateInlineIsInstanceCondPost(A o, boolean cond) {
+        return B.class.isInstance(o) && cond;
+    }
```

Comment:
What about the non‑late version of these methods?

Reply:
It might be useful to have these tests in case the default inlining IR changes in the future.

Reply:
It's intended as a unit test. It's better to catch inlining issue with targeted tests on inlining. From compiler perspective, there's no reason to cover other cases here. There are so many different scenarios how a subtype check can show up in IR. And different scenarios can theoretically fail due to different reasons.